<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>

<body>
  <!--
    这个案例模拟的就是“异步耗时操作本身就返回一个Promise，比如网络请求、本地IO操作”的场景

    这种场景就既能用半async-await也能用全async-await了，此时推荐使用全async-await
  -->
  <script>
    // 封装一个业务层的get请求
    async function getRequest() {
      try {
        // get请求是个异步耗时操作，并且get请求这个异步耗时操作本身就返回一个Promise
        // 这里我们不通过这个异步耗时操作返回的Promise对象.then来监听到它执行成功和失败的时机，然后分别调用resolve和reject，而是用await和try catch来监听
        const value = await fetch("https://www.baidu.com", {
          mode: 'no-cors', // 设置为 no-cors
        });
        return value;
      } catch (error) {
        throw Error(error);
      }
    };

    async function exeEnv() {
      try {
        // 卡在这里等待接口成功的数据
        const value = await getRequest();
        console.log(`get请求成功：${value}`);

        // 这就是在上报成功的数据（会自动包装进一个Promise对象返回）
        return value;
      } catch (error) {
        console.log(`get请求失败：${error}`);

        // 捕获接口失败的错误，并用throw Error上报失败的错误
        throw Error(error);
      }
    };
    exeEnv();
  </script>
</body>

</html>